GVFS/GVFS.Common/GVFSEnlistmentHealthCalculator.cs

```diff
@@ -0,0 +1,258 @@
+        {
+            return path.Substring(directoryTarget.Length);
+        }
+
+        private void IncreaseDictionaryCounterByKey(Dictionary<string, int> countingDictionary, string key)
+        {
+            if (!countingDictionary.TryGetValue(key, out int count))
+            {
+                count = 0;
+            }
+
+            countingDictionary[key] = ++count;
+        }
+
+        private decimal CalculateHealthMetric(int hydratedFileCount, int totalFileCount)
+        {
+            if (totalFileCount == 0)
+            {
+                return 0;
+            }
+
+            return (decimal)hydratedFileCount / (decimal)totalFileCount;
+        }
+
+        public class GVFSEnlistmentPathData
+        {
+            public List<string> GitFolderPaths;
+            public List<string> GitFilePaths;
+            public List<string> PlaceholderFolderPaths;
+            public List<string> PlaceholderFilePaths;
+            public List<string> ModifiedFolderPaths;
+            public List<string> ModifiedFilePaths;
+
+            public GVFSEnlistmentPathData()
+            {
+                this.GitFolderPaths = new List<string>();
+                this.GitFilePaths = new List<string>();
+                this.PlaceholderFolderPaths = new List<string>();
+                this.PlaceholderFilePaths = new List<string>();
+                this.ModifiedFolderPaths = new List<string>();
+                this.ModifiedFilePaths = new List<string>();
+            }
+
+            public void AddSkipWorkTreeFilePaths(List<string> skipWorkTreeFilePaths)
+            {
+                this.ModifiedFilePaths.Union(skipWorkTreeFilePaths);
+            }
+
+            public void NormalizeAllPaths()
+            {
+                this.NormalizePaths(this.GitFolderPaths);
+                this.NormalizePaths(this.GitFilePaths);
+                this.NormalizePaths(this.PlaceholderFolderPaths);
+                this.NormalizePaths(this.PlaceholderFilePaths);
+                this.NormalizePaths(this.ModifiedFolderPaths);
+                this.NormalizePaths(this.ModifiedFilePaths);
+            }
+
+            private void NormalizePaths(List<string> paths)
+            {
+                for (int i = 0; i < paths.Count; i++)
+                {
+                    paths[i] = paths[i].Replace('\\', GVFSConstants.GitPathSeparator);
+                    paths[i] = paths[i].TrimStart(GVFSConstants.GitPathSeparator);
+                    paths[i] = paths[i].TrimEnd(GVFSConstants.GitPathSeparator);
+                }
+            }
+        }
+
+        public class GVFSEnlistmentHealthData
+        {
+            public GVFSEnlistmentHealthData(
+                string targetDirectory,
+                int gitItemsCount,
+                int placeholderCount,
+                int modifiedPathsCount,
+                decimal healthMetric,
+                List<KeyValuePair<string, decimal>> directoryHydrationLevels)

```

Comment:
I personally don't like using something like `KeyValuePair<string, decimal>` because it gives no information about what it represents for example is the `string` the full path, relative path, just the folder name?  I like having a class to represent the data:
```
public class DirectoryHydrationPercent
{
    public string FullPath/RelativePath/FolderName { get; set; }
    public decimal PercentHydrated { get; set; }
}
```

You override `GetHashCode` and `Equals` if you want to use it in a dictionary and then use a `HashSet` instead of `Dictionary` if needed.

Reply:
I created an [issue](https://github.com/microsoft/VFSForGit/issues/1294#issue-460095267) for this so that we can look at it in the future and I can revisit after merging in what is here to be enhanced later.